src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -0,0 +1,721 @@
+            if (parent != null && parent.endIndex == peekIndex) {
+                event = Event.END_CONTAINER;
+            }
+            return true;
+        }
+        reset();
+        int b;
+        if (parent == null) { // Depth 0
+            valuePreHeaderIndex = peekIndex;
+            b = buffer[(int)(peekIndex++)] & SINGLE_BYTE_MASK;
+            if (b == IVM_START_BYTE) {
+                readIvm();
+                return false;
+            }
+        } else {
+            if (nextContainedToken()) {
+                return true;
+            }
+            b = buffer[(int)(peekIndex++)] & SINGLE_BYTE_MASK;
+        }
+        if (readHeader(b, false, valueMarker)) {
+            valueTid = valueMarker.typeId;
+            return true;
+        }
+        valueTid = valueMarker.typeId;
+        return false;
+    }
+
+    @Override
+    public Event nextValue() {
+        event = Event.NEEDS_DATA;
+        while (true) {
+            if (nextToken()) {
+                break;
+            }
+        }
+        return event;
+    }
+
+    @Override
+    public Event fillValue() {
+        event = Event.VALUE_READY;
+        return event;
+    }
+
+    @Override
+    public Event getCurrentEvent() {
+        return event;
+    }
+
+    public int getIonMajorVersion() {
+        return majorVersion;
+    }
+
+    public int getIonMinorVersion() {
+        return minorVersion;
+    }
+
+    public boolean hasAnnotations() {
+        return hasAnnotations;
+    }
+
+    Marker getValueMarker() {
+        return valueMarker;
+    }
+
+    /**
+     * Slices the buffer using the given offset and limit. Slices are treated as if they were at the top level.
+     * @param offset the offset at which the slice will begin.
+     * @param limit the slice's limit.
+     */
+    protected void slice(long offset, long limit) {
+        peekIndex = offset;
+        this.limit = limit;
+        setCheckpointBeforeUnannotatedTypeId();
+        valueMarker.endIndex = -1;
+        event = Event.NEEDS_DATA;
+        valueTid = null;
+        containerIndex = -1; // Slices are treated as if they were at the top level.
+    }
+
+    /**
+     * @return the total number of bytes read since the stream began.
+     */
+    long getTotalOffset() {
+        return valuePreHeaderIndex - startOffset;
+    }
+
+    boolean isByteBacked() {
+        return true;
+    }
+
+    public void registerIvmNotificationConsumer(IvmNotificationConsumer ivmConsumer) {
+        this.ivmConsumer = ivmConsumer;
+    }
+
+    @Override
+    public Event endStream() {
+        if (isAwaitingMoreData()) {
+            throw new IonException("Unexpected EOF.");
+        }
+        return Event.NEEDS_DATA;
+    }
+
+    /**
+     * @return true if the cursor is expecting more data in order to complete a token; otherwise, false.
+     */
+    private boolean isAwaitingMoreData() {
+        return valueMarker.endIndex > limit;
+    }
```

Comment:
If I call `next()` or `stepOut()` on this and it returns `Event.NEEDS_DATA`, is the contract that the only thing the user can do with the `IonCursorBinary` is add more data and then call the same method again? (That is, I add data and call `stepOut()` a second, third, and fourth time until it returns `NEEDS_INSTRUCTION` instead of `NEEDS_DATA`.)

You can imagine calling `stepOut()`, the cursor's event getting set to `NEEDS_DATA`, and then you add more bytes and the cursor quietly sets its event to `NEEDS_INSTRUCTION` once it's satisfied. (That is: the cursor tries to complete the action-in-progress each time it gets new data.)

Whatever the expectation is, I think we should explain that usage in a doc comment.

Reply:
That is the expectation. However, it is not enforced, and is technically not even necessary, which is why I haven't gone out of my way to document a requirement. If the cursor runs out of input before completing the requested task (`next`, `stepOut`, or `fillValue`), it will "remember" the type of operation that was requested (either "fill" or "seek"), and ensure that operation finishes before processing the next request from the user (see [`slowMakeBufferReady`](https://github.com/amazon-ion/ion-java/blob/reentrant-reader-base-pr25/src/com/amazon/ion/impl/IonCursorBinary.java#L963)).

So, let's say the user asks to `stepOut` of the current container, but the cursor runs out of input before reaching the end of that container. It will record that this was a "seek" operation, and that X additional bytes are needed. The user _could_ subsequently call `next` instead of `stepOut`, but before processing `next`, the cursor will ensure that it can complete the "seek(X)" first (or return NEEDS_DATA). In other words, there's no way to "cancel" an operation. I'll add some more documentation about this.

> You can imagine calling stepOut(), the cursor's event getting set to NEEDS_DATA, and then you add more bytes and the cursor quietly sets its event to NEEDS_INSTRUCTION once it's satisfied. (That is: the cursor tries to complete the action-in-progress each time it gets new data.)

I agree that this would be nice. The issue is that the IonCursor does not have an API for "provide more data", unlike, say, ion-python's non-blocking implementation. "Getting data in" is left to a separate abstraction (currently just `InputStream`, but we could support `ByteBuffer` and/or custom abstractions), from which the cursor pulls. Accordingly, the reader will only find out if enough data is available as a side effect of trying to retrieve that much data in response to a user request.

It should be noted that none of this matters when the cursors are wrapped with the IonReader abstraction, which is currently the only way to consume them. That's because there are two IonReader modes:
* non-incremental, where it is an error to encounter an incomplete value, making this question moot
* incremental, where an entire top-level value is buffered *before* it is surfaced to the user. This makes it impossible to encounter a premature end-of-stream during `stepOut`.